be/src/storage/lake/tablet_manager.cpp

```diff
@@ -787,94 +779,96 @@ std::set<int64_t> TabletManager::owned_tablets() {
         }
-        // TODO: wait until all tasks finished.
     }
-    return nullptr;
+    while (num_running.load() > 0) {
+        LOG_EVERY_N(INFO, 10) << "Waiting for GC tasks to finish...";
+        bthread_usleep(/*100ms=*/100 * 1000);
+    }
+
+    return butil::gettimeofday_s() + config::lake_gc_segment_check_interval;
 }
 
-void* segment_gc_trigger(void* arg) {
+void* gc_checker(void* arg) {
     auto tablet_mgr = static_cast<TabletManager*>(arg);
     auto lp = tablet_mgr->location_provider();
     // NOTE: Share the same thread pool with local tablet's clone task.
-    auto thread_pool = ExecEnv::GetInstance()->agent_server()->get_thread_pool(TTaskType::CLONE);
+    int64_t curr = butil::gettimeofday_s();
+    int64_t gc_time[2] = {curr + config::lake_gc_metadata_check_interval,
+                          curr + config::lake_gc_segment_check_interval};
     while (!bthread_stopped(bthread_self())) {
-        // NOTE: When the work load of bthread workers is high, the real sleep interval may be much longer than the
-        // configured value, which is ok now.
-        if (bthread_usleep(config::lake_gc_segment_check_interval * 1000 * 1000) != 0) {
-            continue;
+        int64_t now = butil::gettimeofday_s();
+        int64_t min_gc_time = std::min(gc_time[0], gc_time[1]);
+        if (min_gc_time > now) {
+            // NOTE: When the work load of bthread workers is high, the real sleep interval may be much longer than the
+            // configured value, which is ok now.
+            (void)bthread_usleep((min_gc_time - now) * 1000ULL * 1000ULL);
         }
 
         std::set<std::string> roots;
-        auto st = lp->list_root_locations(&roots);
-
-        LOG_IF(ERROR, !st.ok()) << st;
-
-        for (const auto& root : roots) {
-            // TODO: limit GC concurrency
-            st = thread_pool->submit_func([=]() {
-                auto t1 = std::chrono::steady_clock::now();
-                auto r = segment_gc(root, tablet_mgr);
-                auto t2 = std::chrono::steady_clock::now();
-                auto cost = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - t1).count();
-                if (r.ok()) {
-                    LOG(INFO) << "Finished garbage collection of data for directory " << root << ". cost:" << cost
-                              << "ms";
-                } else {
-                    LOG(WARNING) << "Fail to do garbage collection of data for directory " << root << ". cost:" << cost
-                                 << "ms error:" << r;
-                }
-            });
-            if (!st.ok()) {
-                LOG(WARNING) << "Fail to submit task to threadpool: " << st;
-                break;
-            }
+        (void)lp->list_root_locations(&roots);
+
+        if (min_gc_time == gc_time[0]) {
+            auto master_info = get_master_info();
+            gc_time[0] = metadata_gc(tablet_mgr, roots, master_info.min_active_txn_id);
```

Comment:
specific reason to let `metadata_gc` and `data_gc` return the GC time before return? can gc_checker do it?

Reply:
Updated, PTAL